src/Extension/Mention/Generator/MentionGeneratorInterface.php

```diff
@@ -0,0 +1,25 @@
+<?php
+
+/*
+ * This file is part of the league/commonmark package.
+ *
+ * (c) Colin O'Dell <colinodell@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace League\CommonMark\Extension\Mention\Generator;
+
+use League\CommonMark\Extension\Mention\Mention;
+use League\CommonMark\Inline\Element\AbstractInline;
+
+interface MentionGeneratorInterface
+{
+    /**
+     * @param Mention $mention
+     *
+     * @return AbstractInline|null
+     */
+    public function generateMention(Mention $mention): ?AbstractInline;
```

Comment:
I think you're forgetting why this needed to be a callback in the first place and not some interface to be implemented. It's primarily about future-proofing internal/custom integration implementations.

By simply invoking a callback and passing a single an existing object, it allows the namespace checks to be done during runtime and not bound to any type-hinted parameters/return values or namespace declarations. It also means it (technically) doesn't have to create new elements using internal API namespaces:

```php
<?php
public function createUserMention($mention)
{
    // Immediately return if not passed the supported Mention object.
    // By checking this here, it avoids using hard type-hinted parameters
    // and namespaces which can cause a PHP fatal. It allows for external
    // custom implementations to weather potential internal API changes
    // that may occur over time/newer versions.
    if (!($mention instanceof \League\CommonMark\Extension\Mention\Mention)) {
      return null;
    }
    // Blah, blah.
    $mention->setUrl('...');
    $mention->setLabel('...');
    // No creation of new elements necessary.
}
```

> having this partial "maybe use me maybe don't" object

I get what you're saying here, but the majority of the time it's going to be setting the URL for the mention, probably overriding the label. That's why it's being passed as the primary object: this is the 90% use case of mentions and how it's intended to be used.

If someone decides to take it further and implement something a little more advanced by returning something else (using more of the internal APIs), then that's on them.

The primary goal with this callback should be setting the mention's URL and label.

I'd really prefer that to be a strongly typed object for future-proofing reasons as its much easier to do if/else checks of the existence of namespaces/classes, properties, methods over arbitrary scalar values.

> Implementors are then responsible for adding whatever `Link` or `Text` or whatever child nodes they want and return `true` to indicate success, or return `false` to indicate that it's not a valid mention.

Except that there are really no more "implementers" now that it's all config based now. Unless we want to add another "renderer" class property to the mention config which doesn't really sound like a great idea TBH and I think just makes it infinitely more complex than it really needs to be.

Regardless, if that were to happen it would change the fact that the implementers that are using the advanced callback would likely lose the context of what it is they've returned in the first place (i.e. the user, permissions, whatever else happens during that process).

This is the other reason I extended this from Link/AbstractWebResource/AbstractInline/Node though: it allows the callback to stay within the same context of the services they've used to determine what the mention should be and also give them the ability to set additional data on the mention (or whatever AbstractInline it decides to return):

```php
$mention->data['attributes']['data-tooltip'] = '...';
```

All within the same singular implementation, future-proof implementation.

Reply:
> I think you're forgetting why this needed to be a callback in the first place and not some interface to be implemented. It's primarily about future-proofing internal/custom integration implementations.

You're right, I did forget that.  My apologies!

I still have significant reservations about passing in a partially-constructed `Mention` object where you can't safely call its methods without getting a `TypeError`.

Perhaps instead of passing in a `Mention` and returning `?AbstractInline` we should instead pass in some kind of builder object like this:

```php
final class MentionBuilder
{
    /** @var string */
    private $type;

    /** @var string */
    private $symbol;

    /** @var string */
    private $identifier;

    /** @var string */
    private $label;

    /** @var string|null */
    private $url;
    
    // TODO: Set the first four based on what we parsed (via constructor).

    // TODO: Also provide getters for the first three and getters + setters for the rest.
}
```

We would pass that object into the callback function so users can call `getSymbol()` or `setUrl()` or whatever else they want.  They would then return `true` to indicate success or `false` if no mention should be parsed.

Once the callback returns, we can pull the values out of the `MentionBuilder` to construct the node(s) as needed.

This kind of approach would allow the callbacks to be oblivious to namespaces and types if they desire, while also ensuring this library has an object representation with types that we can use and which isn't only partially constructed.

What are your thoughts on using that type of approach?

Reply:
> I still have significant reservations about passing in a partially-constructed `Mention` object where you can't safely call its methods without getting a `TypeError`.

Where would a TypeError come into play? A simple `instanceof` in the beginning of the callback or any other sort of checks like `property_exists` or `method_exists` on the passed object is sufficient.

This approach provides implementers a way to futureproof their code by allowing BC with any previous objects or limitations the previous object may have. That's why there isn't a type-hint on the `$mention` constructor argument.

It intentionally allows anything to be passed as that first parameter and lets the callback handle compatibility inside itself in a non-fatal manner.

> We would pass that object [`MentionBuilder`] into the callback function so users can call `getSymbol()` or `setUrl()` or whatever else they want. They would then return `true` to indicate success or `false` if no mention should be parsed.
> 
> Once the callback returns, we can pull the values out of the `MentionBuilder` to construct the node(s) as needed.

Ok. So it seems you're on board with passing a single object. Why does it have to be a special "one use only" object though?

Why not just create the object that is intended to be used in the first place? Pass this object off to the callback to provide that missing piece the callback was implemented for in the first place (setting the URL) and then just use that as the child object that's appended.

There are also two other reasons why this will not work:

1. Returning a simple boolean means we'll also lose the ability to return a custom `AbstractInline` objects.
2. No attributes (via data) could be added unless we also add this capability to this `MentionBuilder` object. At which point, we're essentially re-creating all the classes `Mention` already inherits. Why are we reinventing the wheel here by recreating existing code for one-time use?

> This kind of approach would allow the callbacks to be oblivious to namespaces and types if they desire

Check. Which is what the current PR's implementation using the `Mention` object does.

> while also ensuring this library has an object representation with types that we can use

Check. Which again is what the current PR's implementation does; save for that last bit of that paragraph... 

> and which isn't only partially constructed.

Both this PR's `Mention` object and the proposed `MentionBuilder` are only partial constructions.

The reason is: we don't know what the URL is supposed to be. That is why it is a callback in the first place as its entire point of existence is to provide this dynamic URL.

Originally, this callback handled this by returning the URL as a string and assumed it was going to be a `Link` that would be constructed from that.

Since objects are inherently passed by reference, this allows the callback to instead set the URL on the passed object itself. This allows us to simplify the return value expectations of the callback to a singular strongly typed `AbstractInline` which is nullable.

If a URL isn't set, it is essentially the same as returning `null` and saying "this mention isn't a match". If another new `AbstractInline` based object is returned, that is used instead.

All ^ could be summarized in something like:

"Hey, here's the `Mention` object we've created on the assumption you're going to want to use it for 99% of your use-cases. We just need you to set its URL you're supposed to be dynamically generating here, give us a completely new `AbastractInline` object to use instead, or do nothing if this mention isn't a match after all."

Reply:
> Where would a TypeError come into play?

Your `Mention` class doesn't call the parent constructor which sets the `$url`.  Therefore, anyone calling `->getUrl()` in the callback will get a type error because it will attempt to return a `null`:

https://github.com/thephpleague/commonmark/blob/4e4170434782f5addb8d531c3f835e6a3e9957a7/src/Inline/Element/AbstractWebResource.php#L29-L35

This is what I mean by the object being only partially constructed.

(I think our static analysis tools in `master` require the parent constructor to be called, too.)

> This approach provides implementers a way to futureproof their code by allowing BC with any previous objects or limitations the previous object may have. That's why there isn't a type-hint on the $mention constructor argument.

Personally I think the best way to offer compatibility is providing a strongly-typed API and following semver so we can **guarantee** compatibility.  Developers should ideally be writing code like this:

```php
$environment->addExtension(new MentionExtension());
$environment->setConfig([
    'mentions' => [
        'twitter' => [
            'symbol' => '@',
            'regex' => '/^[A-Za-z0-9_]{1,15}(?!\w)/',
            'generator' => new class implements MentionGeneratorInterface {
                // TODO: Typed method goes here
            }
        ],
]);
```

If we did this properly, the interface we implement for 1.5 would remain identical to 2.x.

But I'm also sympathetic to your use case where you want to maximize the chance of compatibility, which is why I'm open to implementing that `CallbackGenerator` which provides that strongly-typed, interfaced class that can wrap the compatibility you want.  I'm happy to be flexible to accommodate that, but at the same time I can't justify introducing other code that would violate the strict typing and everything else that other people would rely on.

That's the main reason why I don't want to introduce a `Mention` class that doesn't follow the same guarantees everything else does.

> Ok. So it seems you're on board with passing a single object. Why does it have to be a special "one use only" object though?

Because the proposed `Mention` object violates the guarantees we provide about type safety.  We therefore need something else that encapsulates that type of information.

> Why not just create the object that is intended to be used in the first place?

That would absolutely be my preference, but to do so within the callback function, which I know you want to avoid.

> Returning a simple boolean means we'll also lose the ability to return a custom AbstractInline objects.

You can't do that without having your custom objects having a `use` statement to the namespace that `AbstractInline` lives in.  If I understand you correctly, you're trying to avoid those types of references, so I don't think this would be an applicable use case for using the simple callback approach - implementing `MentionGeneratorInterface` would be better since you have to reference the namespaces anyway.

> No attributes (via data) could be added unless we also add this capability to this MentionBuilder object. At which point, we're essentially re-creating all the classes Mention already inherits.

That's a fair point, but if we can't use a partially-constructed `Mention` and we can't construct a `new Mention()` from within the callback, then I think it's our only option - otherwise we're back to returning a URL `string`.

----

Ultimately I have no problem providing a callback wrapper and I can see how it would be valuable to simplify the most common operations.  I just can't justify breaking certain promises, expectations, and best practices that others rely on.  I also think that perhaps the scope of this callback function might be too broad - if somebody really wants to return custom node types and the whole works then perhaps they should implement `MentionGeneratorInterface` which lets them do all of that, and we keep the callback approach focused on only the most common use cases.

> "Hey, here's the Mention object we've created on the assumption you're going to want to use it for 99% of your use-cases. We just need you to set its URL you're supposed to be dynamically generating here, give us a completely new AbastractInline object to use instead, or do nothing if this mention isn't a match after all."

I'm essentially proposing this should be:

> "Hey, here's a ~Mention~ MentionBuilder object we've created on the assumption you're going to want to use it for 99% of your use-cases. We just need you to set its URL you're supposed to be dynamically generating here, ~give us a completely new AbastractInline object to use instead,~ or do nothing if this mention isn't a match after all. **If you need more flexibility, we have this nice `MentionGeneratorInterface` you can use instead.**"

Reply:
We continued this discussion in Slack and came up with a game plan. The latest commit addresses the `TypeError` issue. The rest we discussed and are moving forward with the `Mention` object and explicitly requiring the return of the same passed mention object that was passed, after the URL has been set.

This also includes some better documentation regarding different types of callback implementation and when to use them (completely anonymous vs. interfaced/type-hinted).